feat: Hide Child Records for a Nested DocType via User Permissions

marination opened the pull request:
Docs: https://github.com/frappe/erpnext_documentation/pull/238

**Problem:**
- What happens currently is: when a User Permission is added for a nested doctype value, the children records are automatically visible. 
- Child perms are added if descendants exist for a value to restrict.
- E.g. a User Permission is created for user "marica" , restricted to company "Frappe Tech". "Frappe Tech" has a child company "Frappe Apps". "marica" can implicitly view all "Frappe Apps" related records. 
- This isn't very clear to users and this behaviour cannot be overriden using User Perms (descendant perms are added under the hood).

**Solution:**
- In User Permissions show a checkbox called "Exclude Descendants" (we could do better on the naming front) on selection of a nested doctype (Employee, Company, etc.)
![user-perms](https://user-images.githubusercontent.com/25857446/104493954-cda18900-55fb-11eb-8a72-94c29120ddf9.gif)

- On selection of this checkbox, child records wont be visible.
- E.g. (continuing example) If checkbox is enabled "marica" can view all records of "Frappe Tech" only and not its children. If "Allow for Doctype" is added on this perm , let's say "Salary Slip". User "marica" is restricted to company "Frappe Tech" only for Doctype "Salary Slip". "Frappe Apps" records will be hidden.

- Added to Bulk Add/Update user permissions Dialog too (only visible if nested set doctype)
![Screenshot 2021-01-18 at 3 03 35 PM](https://user-images.githubusercontent.com/25857446/104896869-b36dff00-599d-11eb-8666-60c7df97ff8d.png)

marination committed:
feat: Hide Child Records for a Nested DocType via User Permissions

marination committed:
chore: Add 'Exclude Descendants' to add / update perms in bulk dialog

marination committed:
fix: Sider Issues

- Removed unwanted f string in test
- Space after if (client side)

ruchamahabal commented on frappe/core/doctype/user_permission/user_permission.py:
```diff
@@ -194,43 +195,45 @@ def add_user_permissions(data):
 	data = frappe._dict(data)
 
-	d = check_applicable_doc_perm(data.user, data.doctype, data.docname)
+	# get all doctypes on whom this permission os applied
```
```suggestion
	# get all doctypes on whom this permission is applied
```

ruchamahabal commented on frappe/core/doctype/user_permission/user_permission.js:
```diff
@@ -26,11 +26,15 @@ frappe.ui.form.on('User Permission', {
 			() => frappe.set_route('query-report', 'Permitted Documents For User',
 				{ user: frm.doc.user }));
 		frm.trigger('set_applicable_for_constraint');
+		frm.trigger('show_exclude_descendants');
```
```suggestion
		frm.trigger('toggle_exclude_descendants');
```
Maybe, rename this as `toggle_exclude_descendants` instead of `show_exclude_descendants` because it will also hide the checkbox.

ruchamahabal commented on frappe/core/doctype/user_permission/user_permission.js:
```diff
@@ -26,11 +26,15 @@ frappe.ui.form.on('User Permission', {
+			if (frm.doc.for_value) {
+				frm.set_value('for_value', null);
+			}
+			frm.trigger('show_exclude_descendants');
```
```suggestion
			frm.trigger('toggle_exclude_descendants');
```

ruchamahabal commented on frappe/core/doctype/user_permission/user_permission.js:
```diff
@@ -43,6 +47,11 @@ frappe.ui.form.on('User Permission', {
 		}
+	},
+
+	show_exclude_descendants: frm => {
```
```suggestion
	toggle_exclude_descendants: frm => {
```

ruchamahabal reviewed (commented):
LGTM. Working as expected. Not sure but maybe, we can rename "Exclude Descendants" to "Hide Descendants". I think "exclude" will bring in ambiguity as to excluded from what.

ruchamahabal committed:
Merge branch 'develop' into nested-doc-user-perms

marination committed:
chore: Rename 'Exclude Descendants' to 'Hide Descendants'

- Rename 'Exclude Descendants' to 'Hide Descendants'
- Rename js trigger to 'toggle_hide_descendants'

marination commented:
@ruchamahabal done

ruchamahabal commented on frappe/core/doctype/user_permission/user_permission_list.js:
```diff
@@ -74,6 +79,19 @@ frappe.listview_settings['User Permission'] = {
+					},
+					{
+						fieldname: 'hide_descendants',
+						label: __('Exclude Descendants'),
```
```suggestion
						label: __('Hide Descendants'),
```

marination replied:
done

marination committed:
fix: checkbox label in UP List View Dialog

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

Olawale1 commented:
FANTASTIC!!!

surajshetty3416 committed:
Merge branch 'develop' into nested-doc-user-perms
